### PR TITLE
[BEAM-10925] Create interface for SQL Java aggregate function.

### PR DESCRIPTION
**Pull request by ibzib**
R: @amaliujia @kennknowles 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/AggregateFn.java
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/AggregateFn.java:24]: Do you think whether we should start by adding `@Experimental`?

Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/AggregateFn.java:26]: Definitely. Done.

**Review by amaliujia — APPROVED** [review]

**Comment by kennknowles** [comment]
Just checking: the interface for a SQL aggregation function is "can give you a CombineFn"? I understand this is the current status of UDAFs. Is that where we want to leave it? I think the main value of a separate interface (style nit: prefer to separate the interface from this particular implementation) would be a slight decoupling.

(non-blocking, just questions & comments & I'll leave it to others)

**Comment by amaliujia** [comment]
If we don't expose CombineFn in interface, I think what we will end is a similar interface design to current CombineFn: we will ask users to offer implementations `init accumulator`, `add input`, `merge accumulator`, `extract`, then in implementation we will create CombineFn wrapper.

The benefits for this decoupling seems little given CombineFn is a stable API in Beam.

**Comment by amaliujia** [comment]
Also from potential user's requetes, there are existing CombinFn implementation and they want use those as UDAF. This is different case from scalar fn that rewrite CombinFn into our UDAF API may be non-trivial.

**Comment by kennknowles** [comment]
CombineFn is stable, but UDAF is not. One example is that a UDAF has to have a SQL type. Right now this is not represented on the UDAF object but is implied. That might change. Personally, I would make the type part of the UDAF object. If you look at any other SQL product's definition of UDAFs you will see lots of extra metadata ([postgresql](https://www.postgresql.org/docs/9.5/sql-createaggregate.html), [mssql](https://docs.microsoft.com/en-us/sql/t-sql/statements/create-aggregate-transact-sql?view=sql-server-ver15)).

A decoupled interface and trivial wrapper have obvious benefits and no real downside.

**Comment by amaliujia** [comment]
I am not sure if I have understand: 
>One example is that a UDAF has to have a SQL

This interface is the implementation interface of a UDAF, not a definition. Function definition should provide SQL types, but definition in our case is the CREATE FUNCTION statement.

**Review by apilloud — APPROVED** [review]
How do we determine the name of the UDAF? Should there be a name method? Is there another interface for discovering these?

This mostly matches the interface of `registerUdaf`, so LGTM. I agree with Kenn that we will eventually want to offer a version of this interface that doesn't have a dependency on Beam.

**Comment by ibzib** [comment]
> How do we determine the name of the UDAF? Should there be a name method? Is there another interface for discovering these?

I will add another interface to discover these after this PR and #13305 are merged. Something like this:

https://github.com/ibzib/beam/blob/0499c8f08839a9846ecb726b78eef738d21d0052/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/UdfProvider.java

**Comment by ibzib** [comment]
As requested I rewrote this without a dependency on Beam Java core. Right now it is a shameless copy-paste of some stuff from `CombineFn`, since the implementation will still depend on CombineFn anyway. But now we're decoupled at the API level, so we have more flexibility to make whatever changes are needed later (such as Kenn's suggestion to add a SQL type).

> I don't agree in this case that creating a wrapper is trivial. For UDAF/CombeFn, I think it is non-trival work to create the wrapper.

There already is a wrapper that's basically the same as what we're doing here: https://github.com/apache/beam/blob/release-2.25.0/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/agg/AggregationCombineFnAdapter.java

Anyway, I will figure it out :)

File: sdks/java/extensions/sql/udf/src/main/java/org/apache/beam/sdk/extensions/sql/udf/AggregateFn.java
Comment (amaliujia) [inline sdks/java/extensions/sql/udf/src/main/java/org/apache/beam/sdk/extensions/sql/udf/AggregateFn.java:32]: Mark as @experimental?

**Review by amaliujia — APPROVED** [review]
I think there is a experimental annotation is missing. Otherwise LGTM

**Comment by ibzib** [comment]
> I think there is a experimental annotation is missing.

Experimental itself is a part of Beam java core: https://github.com/apache/beam/blob/486d23007dac513f9f90bb08c2a039cd72a7ec94/sdks/java/core/src/main/java/org/apache/beam/sdk/annotations/Experimental.java

I added a note to the javadoc instead.

**Review by apilloud — APPROVED** [review]
LGTM. Make sure you get the changes to build.gradle from #13305...

**Pull request merged**
